File: Dictionaries and Sets/383. Ransom Note.py
from collections import Counter

class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        letters = Counter(magazine)
        
        for c in ransomNote:
            if c not in letters:
                return False
            if letters[c] == 1:
                del letters[c]
            else:
                letters[c] -= 1
        
        return True

# Time: O(n + m)
# Space: O(n)

File: Dictionaries and Sets/383. Ransom Note/383. Ransom Note.js
/**
 * @param {string} ransomNote
 * @param {string} magazine
 * @return {boolean}
 */
var canConstruct = function (ransomNote, magazine) {
  let letters = {};
  for (const m of magazine) {
    if (m in letters) letters[m]++;
    else letters[m] = 1;
  }

  for (const c of ransomNote) {
    if (!(c in letters)) return false;
    if (letters[c] === 1) delete letters[c];
    else letters[c]--;
  }

  return true;
};

// Time: O(n + m)
// Space: O(n)
